### PR TITLE
Custom Queues Merged for 1.11

### PR DESCRIPTION
**Pull request by greezybacon**
So I'm abandoning my previous pull request in favor of a new one. Since custom queues are making their official debut in v1.11 (yay!), I've taken the liberty to start finding the differences between my branch and the upstream branch. This pull request is the start of the conversation to tie up all the loose ends and merge the two ideas.

I think I've got most everything. A couple of the commits showed differences, but the merge conflicts were too extreme for me to figure out. I'll try another couple approaches to make sure we've got everything, but I think this is the most of it.

The only major enhancement here is the concept of creating team-based queues, where the members of a team can share a common queue. 

This also includes a database usage fix to reduce the queries needed to render the queue system by like 90% (From O(n) to O(1) where n is the number of queues visible to a staff member).

### Possibly Outstanding
- [x] Add container queues with no criteria and default sub-queues. This will replace the current "Show assigned tickets on open queue" setting and friends. When the container queue is clicked, the default sub-queue will be shown instead
- [ ] Add description (queue page title) field
- [ ] Add icon selection to custom queue management page
- [ ] Retire CDATA usage and feature
- [ ] Delete custom column (definition)
- [ ] Delete custom sort criteria (definition)
- [ ] Personal queue columns don't need to be translatable
- [ ] Make queue titles translatable

### CODE REVIEW

File: include/class.queue.php [hunk 1]
```diff
@@ -65,8 +70,9 @@ class CustomQueue extends VerySimpleModel {
     const FLAG_INHERIT_DEF_SORT = 0x0040; // Inherit default selected sort
     const FLAG_INHERIT_EXPORT  =  0x0080; // Inherit export fields from parent
 
+    const FLAG_VISIBLE_TEAM =     0x0100; // `staff_id` represents a team_id
 
-    const FLAG_INHERIT_EVERYTHING = 0x158; // Maskf or all INHERIT flags
+    const FLAG_INHERIT_EVERYTHING = 0x1f8; // Maskf or all INHERIT flags
```
Comment (greezybacon) [inline include/class.queue.php]: This was a guess. Should the `INHERIT_EVERYTHING` mask include the new `INHERIT_EXPORT` flag?

File: include/ajax.search.php [hunk 2]
```diff
@@ -153,6 +153,8 @@ function _hashCriteria($criteria, $size=10) {
     function _tryAgain($search, $form=null, $errors=array(), $info=array()) {
         if (!$form)
             $form = $search->getForm();
+        $matches = $search->getSupportedMatches();
```
Comment (greezybacon) [inline include/ajax.search.php]: Is this line necessary? Where is it used?

File: include/class.search.php [hunk 3]
```diff
@@ -844,6 +851,39 @@ function update($vars, &$errors=array()) {
...
+        // If visible to other users, parent queue must also be visible
+        if ($answers['visible_to'] != 'me'
+            && $this->parent
+            # Must be public
+            && (!$this->parent->hasFlag(self::FLAG_PUBLIC)
+            # Or be a queue (which are all public)
+                && !$this->parent->hasFlag(self::FLAG_QUEUE)
+            # Or team must match
+                && !($this->parent->hasFlag(self::FLAG_VISIBLE_TEAM)
+                    && $this->parent->staff_id == $this->staff_id)
```
Comment (greezybacon) [inline include/class.search.php]: Wouldn't `$this->parent->checkAccess($thisstaff)` be better?

File: include/ajax.search.php [hunk 4]
```diff
@@ -391,6 +393,16 @@ function collectQueueCounts($ids=null) {
...
+        $queues = CustomQueue::objects()
+            ->filter(Q::any(array(
+                'flags__hasbit' => CustomQueue::FLAG_PUBLIC,
+                Q::any(array(
+                    'flags__hasbit' => CustomQueue::FLAG_VISIBLE_TEAM,
+                    'team__members__staff_id' => $thisstaff->getId(),
+                )),
+                'staff_id' => $thisstaff->getId(),
+            )));
+
```
Comment (greezybacon) [inline include/ajax.search.php]: Oops. Looks like `$queues` is not used here and might need to be moved to `SavedQueue::ticketsCount`

**Comment by protich** [comment]
> Add container queues with no criteria and default sub-queues. This will replace the current "Show assigned tickets on open queue" setting and friends.

It's already retired as a setting - and now part of **Supplemental Filters** agents can set even on primary queues.

File: include/class.queue.php [hunk 5]
```diff
@@ -920,7 +935,19 @@ function applyDefaultSort($qs) {
...
+            || ($this->hasFlag(self::FLAG_VISIBLE_TEAM)
+                && in_array($this->staff_id, $agent->getTeams()));
+    }
+
+    function checkEditAccess(Staff $agent) {
+        # Either you own it or you're an admin or you're on the team
+        return $agent->isAdmin()
+            || ($agent->getId() == $this->staff_id
+                && !$this->hasFlag(self::FLAG_VISIBLE_TEAM))
+            || (in_array($this->staff_id, $agent->getTeams())
```
Comment (protich) [inline include/class.queue.php]: Once again - this is suspect.

File: include/class.queue.php [hunk 6]
```diff
@@ -1260,6 +1292,53 @@ function save($refetch=false) {
...
+        $criteria = Q::any(array(
+            'flags__hasbit' => self::FLAG_PUBLIC,
+            'flags__hasbit' => static::FLAG_QUEUE,
+            'staff_id' => $staff->getId(),
+        ));
+
+        if (($teams = $staff->getTeams())) {
+            $criteria->add(Q::all(array(
+                'flags__hasbit' => self::FLAG_VISIBLE_TEAM,
+                'staff_id__in' => $teams,
```
Comment (protich) [inline include/class.queue.php]: I have a feeling we need `ost_queue_access` table to truly handle sharing. Think ability to share / restrict queue access to teams, depts and groups as part of sharing.

File: include/class.search.php [hunk 7]
```diff
@@ -672,11 +672,18 @@ function __onload() {
...
+        $criteria = Q::any(array(
             'staff_id' => $agent->getId(),
             'flags__hasbit' => self::FLAG_PUBLIC,
-        )))
-        ->exclude(array('flags__hasbit'=>self::FLAG_QUEUE));
+        ));
+        if ($teams = $agent->getTeams()) {
+            $criteria->add(Q::all(array(
+                'flags__hasbit' => self::FLAG_VISIBLE_TEAM,
+                'staff_id__in' => $teams,
```
Comment (protich) [inline include/class.search.php]: Review

File: include/class.search.php [hunk 8]
```diff
@@ -844,6 +851,39 @@ function update($vars, &$errors=array()) {
...
+        if (!$form->isValid())
+            return false;
+
+        $this->clearFlag(self::FLAG_VISIBLE_TEAM | self::FLAG_PUBLIC);
+        $this->staff_id = $thisstaff->getId();
+        $answers = $form->getClean();
+        switch ($answers['visible_to']) {
+        case 'team':
+            $this->setFlag(self::FLAG_VISIBLE_TEAM);
+            $this->staff_id = $answers['visible_team'];
```
Comment (protich) [inline include/class.search.php]: I **now** see that you're overloading `staff_id` (ignore above comments) -  But I think this goes to show you how confusing it is... adding access table or `team_id` field is a better solution in my opinion.

**Comment by protich** [comment]
> This also includes a database usage fix to reduce the queries needed to render the queue system by like 90% (From O(n) to O(1) where n is the number of queues visible to a staff member).

Anyway you can pull above performance enhancements and other bug fixes for the current implantation as a separate pull request then make the team sharing an independent feature. 

I think doing so will help us cut RC2 sooner and have a separate / longer discussion and perhaps schema changes to address my review comments on share feature.

File: include/class.orm.php [hunk 9]
```diff
@@ -2288,6 +2288,7 @@ static function evaluate($record, $check, $field) {
             'endswith' => function($a, $b) { return iEndsWith($a, $b); },
             'regex' => function($a, $b) { return preg_match("/$a/iu", $b); },
             'hasbit' => function($a, $b) { return ($a & $b) == $b; },
+            'in' => function($a, $b) { return in_array($a, $b); },
```
[the remark this answers is not in the document]
Comment (greezybacon) [reply include/class.orm.php]: These functions should return a boolean result--not an array.

[the remark this answers is not in the document]
Comment (greezybacon) [reply include/class.orm.php]: My bad. I see what you're doing.

**Comment by greezybacon** [comment]
Ok @protich I've rebased this one to remove the sharing support. I'll create a new PR after this is merged.

File: include/class.queue.php [hunk 10]
```diff
@@ -1264,6 +1264,48 @@ function save($refetch=false) {
...
+    static function getHierarchicalQueues(Staff $staff, $pid=0) {
+        $all = static::objects()
+            ->filter(Q::any(array(
+                'flags__hasbit' => self::FLAG_PUBLIC,
+                'flags__hasbit' => static::FLAG_QUEUE,
+                'staff_id' => $staff->getId(),
+                Q::all(array(
+                    'flags__hasbit' => self::FLAG_VISIBLE_TEAM,
+                    'staff_id__in' => $staff->getTeams(),
+                ))
```
Comment (greezybacon) [inline include/class.queue.php]: take this out
